lib/much-result.rb

```diff
@@ -1,4 +1,89 @@
 require "much-result/version"
+require "much-result/item"
 
-module MuchResult
+class MuchResult
+  SUCCESS = "success".freeze
+  FAILURE = "failure".freeze
+
+  Error = Class.new(StandardError)
+
+  def self.success(backtrace: caller, **kargs)
+    new(backtrace: backtrace, **kargs).tap { |result|
+      result.add_item(MuchResult::Item.success(backtrace: backtrace, **kargs))
+    }
+  end
+
+  def self.failure(backtrace: caller, **kargs)
+    new(backtrace: backtrace, **kargs).tap { |result|
+      result.add_item(MuchResult::Item.failure(backtrace: backtrace, **kargs))
+    }
+  end
+
+  def self.for_boolean(value, backtrace: caller, **kargs)
+    new(backtrace: backtrace, **kargs).tap { |result|
+      result.add_item(
+        MuchResult::Item.for_boolean(value, backtrace: backtrace, **kargs)
+      )
+    }
+  end
+
+  attr_reader :description, :backtrace
+
+  def initialize(description: nil, backtrace: caller, **kargs)
+    @description = description
+    @backtrace = backtrace
+    @result_items = []
+
+    set(**kargs)
+  end
+
+  def success?
+    if @success_predicate.nil?
+      @success_predicate =
+        @result_items.reduce(true) { |acc, item| acc && item.success? }
+    end
+
+    @success_predicate
+  end
+
+  def failure?
+    !success?
+  end
+
+  def set(**kargs)
+    @data = ::OpenStruct.new((@data || {}).to_h.merge(**kargs))
+  end
+
+  def add_item(item)
+    @result_items.push(item)
```

Comment:
@kellyredding - Should this reset the `@success_predicate` ivar?

Reply:
@jcredding great eye! And yes, it should (eventually). I chose not to add that logic here b/c I'll need that logic in a coming `.capture` method and I don't know how it will all shake out and I'd prefer to tackle it in that future effort.

Cool by you?

Reply:
I'm gonna assume this is cool. I've addressed it in my next PR, FYI.